File: IAGuide/GuideViewController.m
//
//  GuideViewController.m
//  IAGuide
//
//  Created by Omar Alejel on 11/5/14.
//  Copyright (c) 2014 Omar Alejel. All rights reserved.
//

#import "AppDelegate.h"
#import "DetailTableCell.h" //this was originally for classes but can still use here
#import "TodayManager.h"
#import "GuideViewController.h"
#import "MXLCalendar.h"
#import "MXLCalendarEvent.h"
#import "MXLCalendarManager.h"
#import "ClassBlockViewContainer.h"

@interface GuideViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic) BOOL finishedAnimaton;
@property (nonatomic) BOOL isADay;
@property (nonatomic) BOOL viewAppearedBefore;
@property (nonatomic) ClassBlockViewContainer *scheduleView;
@property (nonatomic) MXLCalendar *eventsCalendar;
@property (nonatomic) NSMutableArray *eventsArray;

@property (nonatomic) UIImage *aImage; //aday
@property (nonatomic) UIImage *bImage; //bday
@property (strong, nonatomic) UIImageView *dayImageView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *tableLoadingView;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIView *calendarHeaderView; // dont mess with this

@property (nonatomic) NSDate *lastOpenedDate;

@end

@implementation GuideViewController

#pragma mark - Initializers
- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"Today" image:[UIImage imageNamed:@"star"] selectedImage:[UIImage imageNamed:@"star_selected"]];
        self.aImage = [UIImage imageNamed:@"aday"];
        self.bImage = [UIImage imageNamed:@"bday"];
    }
    
    return self;
}

#pragma mark - View life cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //register the ClassCell class with the tableview so it can use it in dequeuereusablecell: method
    [self.tableView registerClass:[DetailTableCell class] forCellReuseIdentifier:@"Cell"];
        
    //the position will be set in viewWillAppear
    self.scheduleView = [[ClassBlockViewContainer alloc] init];
    self.scheduleView.layer.cornerRadius = 5;
    self.scheduleView.layer.masksToBounds = YES;
    
    // Do any additional setup after loading the view from its nib.
    //self.dayImageView.image = [UIImage imageNamed:@"bday"];
    self.tableView.backgroundColor = [UIColor clearColor];
    self.tableView.backgroundView.backgroundColor = [UIColor clearColor];
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    self.isADay = [[TodayManager sharedClassManager] todayIsAnADay:[NSDate date]]; //get bool value on whether it is an a day
    UIImage *yesterdayImage;
    if (self.isADay) {
        yesterdayImage = self.bImage; //using opposite image so that i can animate a change
    } else {
        yesterdayImage = self.aImage;
    }
    
    self.dayImageView = [[UIImageView alloc] initWithImage:yesterdayImage];
    self.finishedAnimaton = true; //this will be set to false later
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(appIsActive)
                                                 name:UIApplicationDidBecomeActiveNotification
                                               object:nil];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (!self.viewAppearedBefore) {
        [self setBackgroundGradient];
        [self updateCurrentEvents];
        
        CGPoint origin = self.calendarHeaderView.frame.origin;
        CGSize size = self.calendarHeaderView.frame.size;
        CGRect rect = CGRectMake(self.view.center.x - (size.width/2), origin.y + size.height, size.width, size.width);
        self.dayImageView.frame = rect;
        [self.view insertSubview:self.dayImageView aboveSubview:self.calendarHeaderView];
    
        //add the scheduleView
        CGRect calendarFrame = self.dayImageView.frame;
        CGFloat centerY = ((calendarFrame.origin.y + calendarFrame.size.height) + (self.tableView.frame.origin.y)) / 2;
        self.scheduleView.center = CGPointMake(self.view.center.x, centerY);
        [self.view insertSubview:self.scheduleView aboveSubview:self.tableView];
        self.lastOpenedDate = [NSDate date];
    } else {
        dateFormatter.dateFormat = @"D";
        if ([dateFormatter stringFromDate:self.lastOpenedDate] != [dateFormatter stringFromDate:[NSDate date]]) {
            //day of month not the same; reset tgings that are time dependent
            [self prepareForNewDay];
            self.lastOpenedDate = [NSDate date];
        }
    }
}

- (void)appIsActive {
    dateFormatter.dateFormat = @"D";
    if ([dateFormatter stringFromDate:self.lastOpenedDate] != [dateFormatter stringFromDate:[NSDate date]]) {
        //day of month not the same; reset tgings that are time dependent
        [self prepareForNewDay];
    }
    self.lastOpenedDate = [NSDate date];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    if (!self.viewAppearedBefore) {
        self.viewAppearedBefore = true;
        [self switchDayImage];
    }
}

//set the lgiht blue to blue background gradient
- (void)setBackgroundGradient
{
    //create a gradient for the background
    CAGradientLayer *gradientLayer = [CAGradientLayer layer];
    gradientLayer.frame = self.view.frame;
    
    UIColor *firstColor = [UIColor colorWithRed:105.0/255 green:220.0/255 blue:255.0/255 alpha:1.0];
    UIColor *secondColor = [UIColor colorWithRed:0.0 green:0.17 blue:0.9 alpha:1.0];
    gradientLayer.colors = [NSArray arrayWithObjects:(id)firstColor.CGColor, (id)secondColor.CGColor, nil];
    
    [self.view.layer insertSublayer:gradientLayer atIndex:0];
}

#pragma mark - Current Events

- (void)updateCurrentEvents
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
    MXLCalendarManager *manager = [[MXLCalendarManager alloc] init];
    
        [manager scanICSFileAtRemoteURL:[NSURL URLWithString:@"http://www.iatoday.org/_infrastructure/ICalendarHandler.ashx?Tokens=757278"] withCompletionHandler:^(MXLCalendar *calendar, NSData *nilFileData, NSError *error) {
            
            self.eventsCalendar = calendar;
            
            dispatch_async(dispatch_get_main_queue(), ^{ //make sure this happens instantly
                [self.tableView reloadData];
            });
            if (error) {//no connection, try reading from file
                NSLog(@"Error getting data from online: %@", error.description);
                
                [manager scanICSFileAtLocalPath:[self itemArchivePath] withCompletionHandler:^(MXLCalendar *calendar, NSData *nilFileData, NSError *error) {
                    self.eventsCalendar = calendar;
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self.tableView reloadData];
                    });
                }];
                
            } else {//there is a connection, download for online use
                NSData *calendarFile = [NSData dataWithContentsOfURL:[NSURL URLWithString:@"http://www.iatoday.org/_infrastructure/ICalendarHandler.ashx?Tokens=757278"]];
                
                NSError *writingError;
                [calendarFile writeToFile:[self itemArchivePath] options:0 error:&writingError];
                if (writingError) {
                    NSLog(@"Error writing to file: %@", error);
                }
            }
        }];
    });
    
    [self.tableLoadingView stopAnimating];
}

#pragma mark - A day / B day control

- (void)switchDayImage
{
//    if (self.finishedAnimaton) {
        NSDate *now = [NSDate date];
        //make sure its up to date the moment this is called (for new days)
        self.isADay = [[TodayManager sharedClassManager] todayIsAnADay:now]; //get bool value on whether it is an a day
        
        dateFormatter.dateFormat = @"e";//set format to day of week
        int dayOfWeek = [[dateFormatter stringFromDate:now] intValue];//get number value for weekday
        UIImage *imageToInsert;
        UIImageView *newImageView;
        //
        if (dayOfWeek == 1 || dayOfWeek == 7 || [[TodayManager sharedClassManager] noSchool]) {
            NSDate *testDate = [now dateByAddingTimeInterval:86400];
            for (int i = 0; i < 1000; testDate = [testDate dateByAddingTimeInterval:86400]) {
                //3 if no school that day
                dateFormatter.dateFormat = @"e";//set format to day of week
                int testWeekday = [[dateFormatter stringFromDate:testDate] intValue];//get number value for weekday
                BOOL noBreakBool = [[TodayManager sharedClassManager] dayTypeForDate:testDate] != 3;
                BOOL weekDayBool = (testWeekday != 7) && (testWeekday != 1);
                NSLog(@"Test Weekday: %d, weekdaybool: %@, nobreakbool: %@", testWeekday, weekDayBool ? @"t" : @"f", noBreakBool ? @"t" : @"f");
                if (noBreakBool && weekDayBool) {
                    BOOL nextDayIsADay = [[TodayManager sharedClassManager] todayIsAnADay:testDate];
                    if (nextDayIsADay) {
                        //aday
                        imageToInsert = [UIImage imageNamed:@"nextaday.png"];
                    } else {
                        //bday
                        imageToInsert = [UIImage imageNamed:@"nextbday.png"];
                    }
                    break;//leave the loop
                }
                
                i++;
            }
            
            
        } else {
            //remember that this will handle changing the aday BOOL value for you
            if (self.isADay) {
                //aday
                imageToInsert = self.aImage;
            } else {
                //bday
                imageToInsert = self.bImage;
            }
        }
        
        newImageView = [[UIImageView alloc] initWithImage:imageToInsert];
        newImageView.frame = self.dayImageView.frame;
        [self.view insertSubview:newImageView belowSubview:self.dayImageView];
        self.finishedAnimaton = false;
        CGPoint newCenter = CGPointMake(self.dayImageView.center.x, self.view.frame.size.height + self.dayImageView.frame.size.height);
        
            [UIView animateWithDuration:1.0 delay:0.1 options:UIViewAnimationOptionTransitionNone animations:^{
                self.dayImageView.center = newCenter;
                self.dayImageView.transform = CGAffineTransformMakeRotation(40 * (M_PI/180));
            } completion:^(BOOL finished){
                self.finishedAnimaton = finished;
                [self.dayImageView removeFromSuperview];
                self.dayImageView = nil;
                self.dayImageView = newImageView;
            }];
//    }
}

#pragma mark - TableView

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSUInteger integer = [self.eventsArray count] + 1; //to show cell saying no more
    
    return integer;
}
-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return @"IA Okma Events";
}
- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section
{
    if (view.class == [UITableViewHeaderFooterView class]) {
        UITableViewHeaderFooterView *headerView = (UITableViewHeaderFooterView *)view;
        headerView.backgroundView.backgroundColor = [UIColor grayColor];
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"Cell"];
    cell.textLabel.textColor = [UIColor whiteColor];
    cell.backgroundView.backgroundColor = [UIColor clearColor];
    cell.backgroundColor = [UIColor clearColor];

    if (self.eventsArray.count == indexPath.row) { //last object will be nil
        cell.textLabel.text = @"That's it for the Month!";
        cell.detailTextLabel.text = nil;
        return cell;
    }
    cell.detailTextLabel.textColor = [UIColor whiteColor];
    cell.textLabel.adjustsFontSizeToFitWidth = YES;
    
    MXLCalendarEvent *event = [self.eventsArray objectAtIndex:indexPath.row];
    
    NSString *eventName = [event eventSummary];
    //remove redundancy//
    eventName = [eventName stringByReplacingOccurrencesOfString:@"IA Okma - " withString:@""];
    eventName = [eventName stringByReplacingOccurrencesOfString:@"IA Okma-" withString:@""];
    cell.textLabel.text = eventName;
    
    dateFormatter.dateStyle = NSDateFormatterMediumStyle;
    //set date for event and location of event in one string
    if (event.eventLocation) {
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%@, %@",[dateFormatter stringFromDate:event.eventStartDate], event.eventLocation];
    } else {
        cell.detailTextLabel.text = [dateFormatter stringFromDate:event.eventStartDate];
    }
    
    return  cell;
}

#pragma mark - Setter methods

//called whenever self.eventsCalendar is set
- (void)setEventsCalendar:(MXLCalendar *)eventsCalendar
{
    if (!self.eventsArray) {
        self.eventsArray = [[NSMutableArray alloc] initWithCapacity:10];
    }
    
    NSDate *today = [NSDate date];
    dateFormatter.dateFormat = @"d";
    int dayOfMonth = [[dateFormatter stringFromDate:today] intValue];
    dateFormatter.dateFormat = @"M";//set to month format
    int monthNumber = [[dateFormatter stringFromDate:today] intValue];
    dateFormatter.dateFormat = @"y";//set to year format
    int year = [[dateFormatter stringFromDate:today] intValue];
    
    NSDictionary *daysInMonths = @{@1:@31, @2:@28, @3:@31, @4:@30, @5:@31, @6:@30, @7:@31, @8:@31, @9:@30, @10:@31, @11:@30, @12:@31};
    int daysLeft = [[daysInMonths objectForKey:@(monthNumber)] intValue] - dayOfMonth;
    if (monthNumber == 3 && year % 4 == 0) {
        daysLeft = 29 - dayOfMonth;
    }
    
    for (int i = 0; i < daysLeft; i++) {
        NSDate *newDate = [today dateByAddingTimeInterval:i * 86400];
        dateFormatter.dateFormat = @"d";
        int testDayOfMonth = [[dateFormatter stringFromDate:newDate] intValue];
        dateFormatter.dateFormat = @"M";//set to month format
        int testMonthNumber = [[dateFormatter stringFromDate:newDate] intValue];
        
        for (MXLCalendarEvent *event in eventsCalendar.events) {
            dateFormatter.dateFormat = @"M";//must set month again since reiteration will change dateFormat
            if ([[dateFormatter stringFromDate:event.eventStartDate] intValue] == testMonthNumber) {
                dateFormatter.dateFormat = @"d"; //set to day format
                int f = [[dateFormatter stringFromDate:event.eventStartDate] intValue];
                if (f == testDayOfMonth) {
                    [self.eventsArray addObject:event];
                }
            }
        }
    }
}

//awesome:
- (void)prepareForNewDay {
    dispatch_async(dispatch_get_main_queue(), ^{
        [[TodayManager sharedClassManager] resetForNewDate];
        [self switchDayImage];
        [self updateCurrentEvents];
    });
   
}

#pragma Mark - File functions
//returns the file path for storing the events calendar file
- (NSString *)itemArchivePath
{
    NSArray *documentDirectiories = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentDirectory = [documentDirectiories firstObject];
    
    return [documentDirectory stringByAppendingPathComponent:@"calendar.ics"];
}

@end

File: IAGuide/__GuideViewController.swift
//
//  GuideViewController.swift
//  
//
//  Created by Omar Alejel on 6/18/15.
//
//

import UIKit

class __GuideViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    //booleans
    var finishedAnimation = false
    var isADay = false
    var viewAppearedBefore = false
    
    var scheduleView: ClassBlockViewContainer!
    //school events
    var eventsCalendar: MXLCalendar! {
        didSet {
            updateEventsArray()
        }
    }
    var eventsArray: [MXLCalendarEvent]!    //will hold only current/upcoming
    
    var aImage: UIImage!
    var bImage: UIImage!
    
    var dayImageView: UIImageView!
    @IBOutlet var tableLoadingIndicator: UIActivityIndicatorView!
    @IBOutlet var tableView: UITableView!
    @IBOutlet var calendarHeaderView: UIView!

    //MARK: Initializer
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
//        //set the tab bar item (tab that you use to select the guide controller)
//        let iconUnselected = UIImage(named: "star")
//        let iconSelected = UIImage(named: "star_selected")
//        tabBarItem = UITabBarItem(title: "Today", image: iconUnselected, selectedImage: iconSelected)
//        //set the images so that they are available [make this a computed property later]
//        aImage = UIImage(named: "aday")
//        bImage = UIImage(named: "bday")
    }

    //this is just there because i overrided the init(nib... method
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    //MARK: View Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
//        //set the table cell for events table to detailtablecell
//        tableView.registerClass(DetailTableCell.self, forCellReuseIdentifier: "cell")
//        
//        //setup the view that shows class times
//        scheduleView = ClassBlockViewContainer()
//        scheduleView.layer.cornerRadius = 5//!move to init method of the class
//        scheduleView.layer.masksToBounds = true
//        
//        //tableview setup
//        tableView.backgroundColor = UIColor.clearColor()
//        tableView.backgroundView?.backgroundColor = UIColor.clearColor()
//        tableView.delegate = self
//        tableView.dataSource = self
//        
//        //configure A/B day info and images
//        isADay = TodayManager.sharedClassManager().todayIsAnADay(NSDate())
//        var yesterdayImage: UIImage!
//        if isADay {
//            yesterdayImage = bImage
//        } else {
//            yesterdayImage = aImage
//        }
//        //give the dayImageView an image for yesterdays A/B day value so we can animate it tearing off once the view appears to show the actual A/B day image
//        dayImageView = UIImageView(image: yesterdayImage)
//        finishedAnimation = true//!change the name to animatingDayImage so false is default
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        if !viewAppearedBefore {
            setBackgroundGradient()
//            updateCurrentEvents()
            
//            ///!!!!!fix this section (make it neat and say what it does/....)
//            let origin = calendarHeaderView.frame.origin
//            let size = calendarHeaderView.frame.size
//            //this all needs to change!!!
//            let rect = CGRectMake(view.center.x - size.width / 2, origin.y + size.height, size.width, size.width)
//            dayImageView.frame = rect
//            view.insertSubview(dayImageView, aboveSubview: calendarHeaderView)
//            
//            let calendarFrame = dayImageView.frame
//            let centerY = ((calendarFrame.origin.y + calendarFrame.size.height) + (tableView.frame.origin.y)) / 2
//            scheduleView.center = CGPointMake(view.center.x, centerY)
//            //just make sure its in the right heirarchy
//            view.addSubview(scheduleView)//& = diff!
        }
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        if !viewAppearedBefore {
//            viewAppearedBefore = true
//            switchDayImage()//animate from A->B or vice versa
        }
    }
    
    //MARK: Extra Visual Customization
    //this will set the blue gradient background used throughout the app
    func setBackgroundGradient() {
        let color1 = UIColor(red: 105/255, green: 220/255, blue: 1.0, alpha: 1.0)
        let color2 = UIColor(red: 0.0, green: 0.17, blue: 0.9, alpha: 1.0)
        
        let gradientLayer = CAGradientLayer()
        gradientLayer.frame = view.frame
        gradientLayer.colors = [color1.CGColor, color2.CGColor]
        view.layer.insertSublayer(gradientLayer, atIndex: 0)
    }
    
    func updateCurrentEvents() {
        //go to background thread
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), { () -> Void in
            let calendarManager = MXLCalendarManager()
            let calendarURL = NSURL(string: "https://www.iatoday.org/_infrastructure/ICalendarHandler.ashx?Tokens=757278")
            
            
            calendarManager.scanICSFileAtRemoteURL(calendarURL, withCompletionHandler: { (onlineCalendar, onlineFileData, error) -> Void in
                    //if you cant get data from online, check for saved data
                    if error != nil {
                        print("Error getting data from Online: \(error.description)", terminator: "")
                        calendarManager.scanICSFileAtLocalPath(self.itemArchivePath(), withCompletionHandler: { (fileCalendar, nilFileData, fileError) -> Void in
                            if fileError != nil {
                                print("Error getting data from File: \(fileError.description)", terminator: "")
                            } else if fileError == nil {
                                self.updateTableWithCalendar(fileCalendar)
                            }
                        })
                    } else {
                        //if there was no error in getting from online, save to file
                        //let calendarData = NSData(contentsOfURL: calendarURL!)
                        self.updateTableWithCalendar(onlineCalendar)
                        let data: NSData! = onlineFileData//i did this becuase the was a problem with the compiler
                        data.writeToFile(self.itemArchivePath(), atomically: false)
                }
            })
        })
        
        tableLoadingIndicator.stopAnimating()
    }

    func updateTableWithCalendar(cal: MXLCalendar) {
        //get on the main thread to update the table
        self.eventsCalendar = cal//! might not need
        dispatch_async(dispatch_get_main_queue(), { () -> Void in
            self.tableView.reloadData()
        })
    }
    
    func switchDayImage() {
        if finishedAnimation {
            let now = NSDate()
            isADay = TodayManager.sharedClassManager().todayIsAnADay(now)
            
            dateFormatter.dateFormat = "e"//day of week format (1-7)
            let dayOfWeek = Int(dateFormatter.stringFromDate(now))
            
            var imageToInsert: UIImage!
            
            if dayOfWeek == 1 || dayOfWeek == 7 || TodayManager.sharedClassManager().noSchool() {
                var testDate = now.dateByAddingTimeInterval(86400)
                
                repeat {
                    testDate = testDate.dateByAddingTimeInterval(86400)
                    //get weekday from the date (1-7)
                    let testWeekday = Int(dateFormatter.stringFromDate(testDate))
                    let noHolidayBool = TodayManager.sharedClassManager().dayTypeForDate(testDate) != 3 //3 is when there is a holiday
                    let weekdayBool = (testWeekday != 7) && (testWeekday != 1)
                    
                    if noHolidayBool && weekdayBool{
                        let testIsAnADay = TodayManager.sharedClassManager().todayIsAnADay(testDate)
                        
                        let imageName = testIsAnADay ? "nextaday" : "nextbday"
                        imageToInsert = UIImage(named: imageName)
                        break
                    }
                } while (true);
            } else {
                imageToInsert = isADay ? aImage : bImage
            }
            
            let newImageView = UIImageView(image: imageToInsert)
            newImageView.frame = dayImageView.frame
            view.insertSubview(newImageView, belowSubview: dayImageView)
            finishedAnimation = false
            let newCenter = CGPointMake(view.center.x, view.frame.size.height + dayImageView.frame.size.height)
            
            UIView.animateWithDuration(1.0, delay: 0.1, options: UIViewAnimationOptions.CurveEaseIn, animations: { () -> Void in
                    self.dayImageView.center = newCenter
                    self.dayImageView.transform = CGAffineTransformMakeRotation(CGFloat(40 * (M_PI * 180)))
                }, completion: { (done) -> Void in
                    self.finishedAnimation = true
                    self.dayImageView.removeFromSuperview()
                    self.dayImageView = nil
                    self.dayImageView = newImageView
            })
        }
    }
    
    //MARK: TableView
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        //add 1 to number of events so we can have a "No More!" cell
        if let array = eventsArray {
            return array.count
        } else {
            return 0
        }
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "IA Okma Events"
    }
    
    func tableView(tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        if let headerView = view as? UITableViewHeaderFooterView {
            headerView.backgroundView?.backgroundColor = UIColor.grayColor()
        }
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("Cell", forIndexPath: indexPath) 
        //setup colors
        cell.textLabel!.textColor = UIColor.whiteColor()
        cell.detailTextLabel!.textColor = UIColor.whiteColor()
        cell.backgroundView!.backgroundColor = UIColor.clearColor()
        cell.backgroundColor = UIColor.clearColor()
        
        cell.textLabel?.adjustsFontSizeToFitWidth = true
        
        if eventsArray.count == indexPath.row {
            cell.textLabel?.text = "That's it for the Week!"
            cell.detailTextLabel?.text = ""
            return cell
        }
        
        let schoolEvent = eventsArray[indexPath.row]
        var eventName = schoolEvent.eventSummary
        eventName = eventName.stringByReplacingOccurrencesOfString("IA Okma - ", withString: "")
        eventName = eventName.stringByReplacingOccurrencesOfString("IA Okma-", withString:"")
        cell.textLabel?.text = eventName
        
        let eventStart = schoolEvent.eventStartDate
        let eventLocation = schoolEvent.eventLocation
        
        dateFormatter.dateStyle = NSDateFormatterStyle.MediumStyle
        
        if schoolEvent.eventLocation != nil {
            cell.detailTextLabel?.text = "\(dateFormatter.stringFromDate(eventStart)), \(eventLocation)"
        } else {
            cell.detailTextLabel?.text = "\(dateFormatter.stringFromDate(eventStart))"
        }
        
        return cell
    }
    
    func updateEventsArray() {  //when the events calendar is updated, filter the data
        var date = NSDate()
        for _ in 1...7 {
            date = date.dateByAddingTimeInterval(86400)
            dateFormatter.dateFormat = "d"//day of month format
            let dayOfMonth = Int(dateFormatter.stringFromDate(date))
            dateFormatter.dateFormat = "M"//month format
            let monthNumber = Int(dateFormatter.stringFromDate(date))
            for event in eventsCalendar.events {
                if let eventStart = event.eventStartDate {
                    dateFormatter.dateFormat = "M"//reset to month format
                    if Int(dateFormatter.stringFromDate(eventStart)) == monthNumber {
                        dateFormatter.dateFormat = "d"//day format
                        let f = Int(dateFormatter.stringFromDate(eventStart))
                        if f == dayOfMonth {
                            eventsArray.append(event as! MXLCalendarEvent)
                        }
                    }
                }
            }
        }
    }
    
    func prepareForNewDay() {
        switchDayImage()
        updateCurrentEvents()
    }
    
    func itemArchivePath() -> String {
        let docDirectories = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true)
        let directory = docDirectories.first! as String
        return directory + "/calendar.ics"
    }

}
